### PR TITLE
Speed up debian build process

### LINKED ISSUE #2939: Parallelize deb package builds
# Feature request

## Description

Running `make build-debs` now takes around 20m to complete, given how many Debian packages we're building. Since we switched to containers for the build process, and we're heavily reusing build logic in serial, we're in a good position to leverage the `builder` Molecule scenario to spin up containers concurrently and build Debian packages in parallel.

Specifically, we should be able to consolidate:

* Five (5) invocations of the `build-generic-pkg` role
* Two (2) invocations of the `build-ossec-deb-pkg` role

No changes to the `build-securedrop-app-code-deb-pkg`, which must run separately, due to the unique logic. The minor customizations at the play level in the `build-deb-pkgs.yml` playbook can be ported to Molecule-specific host vars to accommodate parallel execution.

Whether the change is actually a timesaver bears further testing, but I'm confident we can build more quickly than 20m. (If anyone has substantially different build times, please comment; my workstation often runs several virtualized environments, so that's certainly a factor.)

## User Stories
As a developer, I want to test changes quickly and effectively, and not spend a significant portion of my day waiting for build artifacts to validate changes.

### PR DESCRIPTION
**Pull request by msheiny**
## Status

Ready for Review

## Description of Changes

Fixes #2939 . 
Partially addresses #3237  - moves all apt dependencies into docker for caching

Changes proposed in this pull request:

This PR parallelizes the debian build logic and moves all the apt packaging logic into docker (and out of ansible). 

## Testing

How should the reviewer test this PR?

* Run `time make build-debs` - confirm you receive no errors. The timing here will be a first run when docker has to build the image completely.
* Run `time make build-debs` again. This time the docker image doesnt have to get rebuilt and you should see a dramatic time change.
* To compare with existing switch back to `develop` and run `time make build-debs`

If you want to clean out all these docker images first you can run the following selective nuke command `docker image prune -a -f --filter "label=image_name=trusty-sd-builder-app"`

## Deployment

Any special considerations for deployment?

This code re-mixes the build process a bit and removes some tests. Give careful review and feedback to that when reviewing this PR.

We want to make sure that there are no regression in the build process so I would say that during QA we should be looking out for any weird behavior during install or involving missing files.

### CODE REVIEW

**Comment by msheiny** [comment]
okay did some time comparisons specifically looking at CI jobs:

```
Previous build logic - ~10 min
New logic CI (from scratch) - ~8 min
New logic CI (from cache) - 5:32 
```

99% of the time we will be building from cache unless the dockerfile is updated in a particular PR. For some reason CI is way way faster than when I attempt to build locally from scratch (upwards of 20 min). I wonder if that is related to grsec or a docker version ... hmm...

**Comment by msheiny** [comment]
> Is there a particular reason why it needs to be done with this new build strategy?

If you don't pin the versions, we need another way to detect when there are package updates and tell docker to rebuild the underlying image. Otherwise someone could be building with super old packages regardless.

**Comment by msheiny** [comment]
> As a compromise, perhaps CI could check for security updates for installed packages (for example apt-get -s dist-upgrade |grep "^Inst" |grep -i trusty-security)? While imperfect, it would be an improvement.

hah . sorry, I totally ignored your second comment. yeah im fine with appending some kind of CI checker here. to recap convo out of band .. lets discuss... out of band? heh

File: molecule/builder/Dockerfile [hunk 1]
```diff
@@ -6,8 +6,25 @@ ARG NAME
...
 LABEL image_name="$NAME"
 
-# Build dependencies must not be added here, they are listed
-# in install_files/ansible-base/group_vars/all/securedrop
-RUN apt-get -y update && apt-get install -y python rsync sudo bash aptitude && \
-            aptitude -y dist-upgrade && \
-            apt-get clean
+# Non-version specific tooling required
+# Followed by pegged dependencies
+RUN apt-get -y update && apt-get install -y make unzip python python-pip rsync sudo bash aptitude paxctl \
```
Comment (heartsucker) [inline molecule/builder/Dockerfile]: Is there a reason you're not pinning everything?

File: .circleci/config.yml [hunk 2]
```diff
@@ -154,10 +159,28 @@ jobs:
 
       - setup_remote_docker
 
+      - run:
+          name: Load image layer cache
+          command: |
+            set +o pipefail
+            docker load -i /caches/deb-builder.tar |true
```
Comment (heartsucker) [inline .circleci/config.yml]: Why `pipefail` to `true`. Won't that just exit the same as if you ran the command directly?

File: .circleci/config.yml [hunk 3]
```diff
@@ -163,7 +163,7 @@ jobs:
           name: Load image layer cache
           command: |
             set +o pipefail
-            docker load -i /caches/deb-builder.tar |true
+            docker load -i /caches/deb-builder.tar.gz |true
```
Comment (heartsucker) [inline .circleci/config.yml]: Same here with `pipefail` to `true`.

File: .circleci/config.yml [hunk 4]
```diff
@@ -154,10 +159,28 @@ jobs:
 
       - setup_remote_docker
 
+      - run:
+          name: Load image layer cache
+          command: |
+            set +o pipefail
```
Comment (heartsucker) [inline .circleci/config.yml]: Aaaaaand here too.

File: .circleci/config.yml (hunk 2 above)
Comment (msheiny) [reply .circleci/config.yml]: hah honestly i just copied and pasted the logic that was there before that wasnt written by me. Im not opposed to tearing it out as I dont have any justification for it being there.

File: molecule/builder/Dockerfile (hunk 1 above)
Comment (msheiny) [reply molecule/builder/Dockerfile]: i didnt think it made sense to peg some of these auxiliary tools .... buttt as with `pipefail` i dont have a good argument on why we shouldn't pin here as well.

**Comment by conorsch** [comment]
Clear improvement performance-wise: 

```
First build on speedy-debs: 16m46s
Second build on speedy-debs: 8m56s
Third build on speedy-debs: 8m44s

First build on develop: 12m33s
Second build on develop: 14m54s
```

The pinning strategy definitely needs to be refined. The suggestions by @emkll for CI checks are great. Would be willing to merge this work without pinning, but as we already have solid discussion on that end, let's try to revise accordingly. 

Stepping away, since we need a rebase, and there are some outstanding proposals to implement here.

**Comment by msheiny** [comment]
@conorsch / @emkll  / @heartsucker  - i did some changes on the container handling to hopefully address security concerns brought up during this review. Particularly by building and pushing to our quay.io account for snapshot-ing instead of relying on version control pegs (which are annoying to maintain). I also added a test to see if there are security updates to that base container... im not 100% that we should have that check run on every PR... but I think its the quickest compromise I could get in for now.

File: molecule/builder/Dockerfile [hunk 5]
```diff
@@ -1,13 +1,33 @@
-ARG IMAGE
-FROM $IMAGE
+# April 28th, 2018
```
Comment (heartsucker) [inline molecule/builder/Dockerfile]: Nit: Can you include the version you pulled to get this sha? E.g., `trusty-2018-04-28`. I find this makes it easier to track things down when I have to revisit dockerfiles.

File: molecule/builder/Dockerfile [hunk 6]
```diff
@@ -1,13 +1,33 @@
...
+        python-pip \
+        rsync \
+        ruby \
+        secure-delete \
+        sqlite \
+        unzip \
+        && apt-get clean \
+        && rm -rf /var/lib/apt/lists/*
+
+RUN paxctl -cm /usr/bin/python2.7
```
Comment (heartsucker) [inline molecule/builder/Dockerfile]: Nit: You could make these all one `RUN` command to reduce the number of layers that are saved to the dev's disk.

File: molecule/builder/create.yml [hunk 7]
```diff
@@ -8,38 +8,14 @@
     molecule_ephemeral_directory: "{{ lookup('env', 'MOLECULE_EPHEMERAL_DIRECTORY') }}"
     molecule_scenario_directory: "{{ lookup('env', 'MOLECULE_SCENARIO_DIRECTORY') }}"
     molecule_yml: "{{ lookup('file', molecule_file) | from_yaml }}"
+    image_hash: e9f1683b1fe756751e3ec6481860414988eceaa25663e2e98dd5223e6c3b2856
```
Comment (heartsucker) [inline molecule/builder/create.yml]: Where does this hash come from? Can you make a note so it looks less like a magic number.

**Comment by heartsucker** [comment]
Tested with

```
virtualenv -p `which python2` venv
source venv/bin/activate
pip install -r securedrop/requirements/develop-requirements.txt
make build-deb
vagrant up /provision/
```

and then doing one full submission with the source via Tor Browser.

**Pull request merged**